feat: upgrade to webpack5

pi0 opened the pull request:
## Changelog and Links

-  https://github.com/webpack/changelog-v5/blob/master/README.md
- https://github.com/webpack/webpack/issues/8537

## Missing / Delegated

- Arrow syntax is not transpiled for all modes
- ExtractCSS disabled
- Deprecate warnings (for 3rd party WP plugins)
- Font preload test 
- Module Federation investigation
- `@nuxt/components`

## Checklist

- [x] Pin `webpack` version to v5-
- [x] Pin `html-webpack-plugin` to v4-
  - [x] replace `dependency` chunk sort mode by `auto`
- [x] `optimization.splitChunks name: true` removed: Automatic names are no longer supported
- [x] Prefer using `chunkIds: "named"` and `idHint` (default value in development mode is `named`)
- [x] `compilation.hooks.normalModuleLoader` was moved to `NormalModule.getCompilationHooks(compilation).loader`
  - [x] `vue-loader/lib/plugin.js` PR: https://github.com/vuejs/vue-loader/pull/1469
  - [x] `extract-css-chunks-webpack-plugin` (faceyspacey/extract-css-chunks-webpack-plugin#135)
- [x] `[hash]` is now `[fullhash]`
   - [x] Consider using `[chunkhash]` or `[contenthash]`
   - [x] nuxt webpack
   - [x] loader-utils/lib/interpolateName.js
   - [x] webpack/lib/EvalSourceMapDevToolModuleTemplatePlugin.js
- [ ] `mini-css-extract-plugin` and `extract-css-chunks-webpack-plugin` are incompatible (RuntimeModules change) (MainTemplate.hooks.requireEnsure removed)
- [x] `fork-ts-checker-webpack-plugin` is incompatible (frozen hooks)
- [x] `cache-loader` is no longer needed
- [x] `hard-source-webpack-plugin` is obsolete
- [ ] `webpack-node-externals` PR: https://github.com/liady/webpack-node-externals/pull/59 Mirgrated to local for now
- [x] `webpack-dev-middleware` PR: https://github.com/webpack/webpack-dev-middleware/pull/459
- [x] `last-call-webpack-plugin`: dependency of `optimize-css-assets-webpack-plugin`
- [ ] ensure devtool is always matching `/^(inline-|hidden-|eval-)?(nosources-)?(cheap-(module-)?)?source-map$/`
- [x] Update AsyncMFS to use `memfs`

pi0 committed:
pin webpack to 5.0.0-alpha.0

pi0 committed:
update commons chunkGroup name

pi0 committed:
pin html-webpack-plugin to 4.0.0-alpha

pi0 commented:
Tip: Trace deprecation warnings via `node -r esm --trace-warnings --trace-deprecation ./node_modules/.bin/nuxt examples/hello-world`

aldarund commented:
too early :)

pi0 committed:
hash ~> fullhash

pi0 committed:
use auto chunk sorting for HTMLPlugin

pi0 committed:
update extract-css fixture to not use name: true

pi0 committed:
Merge branch 'dev' into feat/webpack5

manniL commented:
@pi0 I guess `HardSourceWebpackPlugin` is obsolete then as well? (PR #4527)

ScriptedAlchemy commented:
Appreciate the work you are doing on Extract CSS Chunks!

clarkdo committed:
feat: upgrade html-webpack-plugin to v4

clarkdo committed:
refactor: downgrade html-webpack-plugin since resolutions is used

clarkdo committed:
refactor: upgrade webpack to alpha.3

clarkdo committed:
fix: lint error

clarkdo committed:
feat: use persistent caching instead of cache-loader

clarkdo committed:
feat: compilers now need to be closed after being used

clarkdo committed:
feat: remove splitChunks.name

pi0 committed:
Merge branch 'dev' into feat/webpack5

clarkdo committed:
feat: remove hard-source-webpack-plugin in favour of options.cache

clarkdo committed:
feat: remove webpack-node-externals for now

ScriptedAlchemy reviewed (commented):
How’s the stability? And its in house css loading. Does css or sass hot reload?

manniL commented on packages/webpack/src/config/base.js:
```diff
@@ -127,6 +126,18 @@ export default class WebpackBaseConfig {
   }
 
+  cache() {
+    if (this.options.build.cache) {
```
Maybe it's worth inverting this condition here and use an early return? :)

clarkdo replied:
Yes, that will be better

clarkdo committed:
refactor: make cache early return

clarkdo committed:
Merge remote-tracking branch 'origin/dev' into feat/webpack5

clarkdo committed:
fix: line error

clarkdo committed:
feat: upgrade fork-ts-checker-webpack-plugin

clarkdo committed:
chore: update yarn.lock

pi0 committed:
Merge branch 'dev' into feat/webpack5

Commit:
fix yarn.lock

Commit:
fix yarn.lock

Commit:
fix: directly upgrade packages

pi0 committed:
Merge branch 'dev' into feat/webpack5

pi0 committed:
fix yarn.lock

clarkdo committed:
chore: webpack@^5.0.0-alpha.9

clarkdo committed:
Merge remote-tracking branch 'origin' into feat/webpack5

clarkdo committed:
chore: fork-ts-checker-webpack-plugin 1.0.0-alpha.6

pi0 committed:
Merge branch 'dev' into feat/webpack5

pi0 committed:
Merge branch 'dev' into feat/webpack5

Commit:
recreate yarn.lock

Commit:
remove futureEmitAssets

Commit:
fix: temporary add wp5

Commit:
disable warn :)

pi0 committed:
Merge branch 'dev' into feat/webpack5

pi0 committed:
Merge branch 'dev' into feat/webpack5

pi0 committed:
Merge branch 'dev' into feat/webpack5

pi0 committed:
Merge branch 'dev' into feat/webpack5

clarkdo committed:
Merge branch 'next' into feat/webpack5

clarkdo committed:
fix lint error

clarkdo committed:
upgrade html plugin

clarkdo committed:
make fileName return func possible

clarkdo committed:
add nuxt config files as additional code dependencies of the build
webpack will use an hash of these items and all dependencies to invalidate the filesystem cache

pi0 committed:
rebase with dev and update deps

pi0 committed:
update package.json(s)

pi0 committed:
Merge branch 'dev' into feat/webpack5

pi0 committed:
read buildOptions from context

pi0 committed:
update devtool

clarkdo committed:
Merge branch 'dev' into feat/webpack5

ScriptedAlchemy commented:
For the removal of requireEnsure - this looks like what it became. https://github.com/webpack/webpack/blob/master/lib/async-modules/AwaitDependenciesInitFragment.js

MainTemplate will be removed in v6

Commit:
mergr with dev + webpack 5-beta.13

Commit:
chore: update some webpack deps

Commit:
update snapshot

Commit:
update snapshot

pi0 committed:
Merge branch 'next' into feat/webpack5

Commit:
update yarn.lock

Commit:
remove memory-fs from ignoreDeps

Commit:
Merge branch 'next' into feat/webpack5

Commit:
chore(deps): update file-loader and url-loader

Commit:
update vue plugin target validation

Commit:
fix(modern): push safariNoModuleFix to data.bodyTags

Commit:
temporary disable extract-css tests

Commit:
update webpack config snapshot

Commit:
disable extract css related tests with todo

Commit:
chore: update html-webpack-plugin to stable and comment out wp5-fix

Commit:
use tap/tapPromise instead of tapAsync/cb

ScriptedAlchemy commented:
Also, I wrote this feature into Webpack 5 called Module Federation @manniL - I’ve rebuilt next.js in order to implement this, since this project is the Vue equivalent, and likely face the “how do we do micro-frontends, with shared resources and no page reloads, along with how do we to the same in SSR as well?” 

This might be of use: https://link.medium.com/i7cpPsXJ64

pi0 committed:
Merge branch 'next' into feat/webpack5

Commit:
Merge branch 'next' into feat/webpack5

Commit:
fix merge issues

Commit:
fix: pass undocumenteed fs option to webpackDevMiddleware

Commit:
test: fix with-config

- font preload TODO
- process.env ~> runtimeConfig

Commit:
fix: reject in case of faild build for webpack-dev

Commit:
test: fix basic.dev

Commit:
test: commons chunk name

Commit:
test: update modern.server

Commit:
test: disable arrow check for now for modern

Commit:
test: fix arrow matcher

clarkdo commented on packages/webpack/src/plugins/wp5-fix.js:
```diff
@@ -0,0 +1,12 @@
+module.exports = class WP5Fix {
```
Can be removed?

pi0 replied:
Yes actually it was to hide deprecated warnings but we should indeed try to update any problematic 3rd party plugin instead of hiding :) Will remove

pi0 replied:
Done

clarkdo commented on packages/webpack/src/config/client.js:
```diff
@@ -74,7 +74,7 @@ export default class WebpackClientConfig extends WebpackBaseConfig {
         chunks: 'all',
         priority: 10,
-        name: true
+        name: 'commons'
```
Will this affect  hash name in prod?

pi0 replied:
>  optimization.splitChunks name: true removed: Automatic names are no longer supported

I think was related to this

clarkdo replied:
WP5 should give name by default, what’s the name on dev/prod if we remove name: commons

pi0 replied:
![image](https://user-images.githubusercontent.com/5158436/81743868-a7db4780-94a2-11ea-975e-74b42db05f74.png)

@clarkdo seems ok. It is like a hint

clarkdo replied:
Great 👍

pi0 replied:
For reference: Removing explicit name, make it hard for bundle analyzer:

![image](https://user-images.githubusercontent.com/5158436/81744946-313f4980-94a4-11ea-991d-26b446718d08.png)

Commit:
chore: remove wp5-fix plugin

ScriptedAlchemy commented:
I’m thrilled to see you’ve added module Federation investigation!
Get hold of me when you start looking into it. I’m supporting angular’s core team as they upgrade to module federation. 

Happy to help or provide information

clarkdo commented:
@ScriptedAlchemy Thanks, it would be great that we can get your help when we start.

ScriptedAlchemy commented:
Get me on twitter and we can set up a zoom call. We should be cutting beta.17 which triples the capabilities of module federation 

If you’ve got time to waste. Here’s a intro to advance api now in master. This only covers shared and sharing. There’s at least 2 more videos that need to be made for the other things we did. https://youtu.be/9F7GoDMrzeQ

pi0 committed:
Merge branch 'next' into feat/webpack5

pi0 committed:
fix typo

Commit:
Merge branch 'next' into feat/webpack5

Commit:
Merge branch 'next' into feat/webpack5

Commit:
chore: update yarn.lock

Commit:
fix: disable @nuxt/components because of webpack incompatibility

Commit:
feat: use memfs

Commit:
test: update tests for components
